test/prob/utility.hpp

```diff
@@ -54,13 +54,69 @@ std::ostream& operator<<(std::ostream& os, const vector<var>& param) {
 }  // namespace std
 
 // ------------------------------------------------------------
+
+template <typename T>
+T get_param(const vector<double>& params, const size_t n) {
+  T param = 0;
+  if (n < params.size())
+    param = params[n];
+  return param;
+}
+
+template <>
+empty get_param<empty>(const vector<double>& /*params*/, const size_t /*n*/) {
+  return empty();
+}
+
+template <>
+fvar<double> get_param<fvar<double>>(const vector<double>& params,
+                                     const size_t n) {
+  fvar<double> param = 0;
+  if (n < params.size()) {
+    param = params[n];
+    param.d_ = 1.0;
+  }
+  return param;
+}
+template <>
+fvar<var> get_param<fvar<var>>(const vector<double>& params, const size_t n) {
+  fvar<var> param = 0;
+  if (n < params.size()) {
+    param = params[n];
+    param.d_ = 1.0;
+  }
+  return param;
+}
+template <>
+fvar<fvar<double>> get_param<fvar<fvar<double>>>(const vector<double>& params,
+                                                 const size_t n) {
+  fvar<fvar<double>> param = 0;
+  if (n < params.size()) {
+    param = params[n];
+    param.d_ = 1.0;
+  }
+  return param;
+}
+template <>
+fvar<fvar<var>> get_param<fvar<fvar<var>>>(const vector<double>& params,
+                                           const size_t n) {
+  fvar<fvar<var>> param = 0;
+  if (n < params.size()) {
+    param = params[n];
+    param.d_ = 1.0;
+  }
+  return param;
+}
+
+// ------------------------------------------------------------
+
 // default template handles Eigen::Matrix
 template <typename T>
 T get_params(const vector<vector<double>>& parameters, const size_t p) {
   T param(parameters.size());
   for (size_t n = 0; n < parameters.size(); n++)
     if (p < parameters[0].size())
-      param(n) = parameters[n][p];
+      param(n) = get_param<stan::scalar_type_t<T>>(parameters[n], p);
```

Comment:
Why/where do you need this change?

Reply:
This is important for forward mode autodiff when we initialize a vector vs. when we initialize scalars.

If the test scalar type is `fvar<var>`, then the scalars get initialized with `get_param`.

This looks like:

```
param = params[n];
param.d_ = 1.0;
```

Previously, when initializing vectors, the initialization looked like:

```
param(n) = parameters[n][p];
```

I think that leaves the `.d_` initialized to zero, which means forward mode autodiff tests fail. Anyway, now we just use the same code to initialize scalars as to initialize each element of a vector so we don't have these problems.

Reply:
Hmm, do you know why were forward tests passing than before this PR? Actually this PR seems like it fixes quite a number of serious bugs. Any idea why nothing failed before?

Reply:
I'm not sure, and give the seriousness of the bugs I kinda worry there's more.

Reply:
Well, I know there's more problems lol. There will be a follow up pull request to this that tests that: `cdf([a, b, c, d])` evaluates to the same thing as `cdf(a) * cdf(b) * cdf(c) * cdf(d)`. Lack of those tests caused a bug previously. I also know in the other branch where I put this all together there were problems with discrete_range which didn't show up here. Not sure what the deal is. Maybe adding that other test will uncover them again so I can put the fixes back in?